petastorm/spark/spark_dataset_converter.py

```diff
@@ -1,94 +1,102 @@
-    :param dir_path: the directory for the saved parquet file, could be local, hdfs, dbfs, ...
-    :return: the path of the saved parquet file
+    If not, cache the df into the cache_dir in parquet format and return the cache file path.
+    :param df:        A :class:`DataFrame` object.
+    :param cache_dir: A string denoting the directory for the saved parquet file.
+    :return:          A string denoting the path of the saved parquet file.
     """
-    # Todo: add cache management
+    uuid_str = str(uuid.uuid4())
+    save_to_dir = os.path.join(cache_dir, uuid_str)
     df.write.mode("overwrite") \
-        .option("parquet.block.size", 1024 * 1024) \
-        .parquet(CACHE_DIR)
+        .option("parquet.block.size", row_group_size) \
+        .parquet(save_to_dir)
 
     # remove _xxx files, which will break `pyarrow.parquet` loading
-    underscore_files = [f for f in os.listdir(dir_path) if f.startswith("_")]
+    underscore_files = [f for f in os.listdir(save_to_dir) if f.startswith("_")]
     for f in underscore_files:
-        os.remove(os.path.join(dir_path, f))
-    return CACHE_DIR
+        os.remove(os.path.join(save_to_dir, f))
+    return save_to_dir
 
 
-def make_spark_converter(df: DataFrame) -> SparkDatasetConverter:
+def make_spark_converter(df, cache_dir=None, row_group_size=ROW_GROUP_SIZE):
     """
-    This class will check whether the df is cached.
-    If so, it will use the existing cache file path to construct a SparkDatasetConverter.
-    If not, Materialize the df into the configured cache dir in parquet format and use the cache file path to
-    construct a SparkDatasetConverter.
-    :param df: The DataFrame to materialize.
-    :return: a SparkDatasetConverter
+    Convert a spark dataframe into a :class:`SparkDatasetConverter` object. It will materialize
+    a spark dataframe to a `cache_dir` or a default cache directory.
+    The returned `SparkDatasetConverter` object will hold the materialized dataframe, and
+    can be used to make one or more tensorflow datasets or torch dataloaders.
+
+    :param df:        The :class:`DataFrame` object to be converted.
+    :param cache_dir: A string denoting the parent directory to store intermediate files.
+                      Default None, it will fallback to the spark config
+                      "spark.petastorm.converter.default.cache.dir".
+                      If the spark config is empty, it will fallback to DEFAULT_CACHE_DIR.
+
+    :return: a :class:`SparkDatasetConverter` object that holds the materialized dataframe and
+            can be used to make one or more tensorflow datasets or torch dataloaders.
     """
-    cache_file_path = _cache_df_or_retrieve_cache_path(df, CACHE_DIR)
-    return SparkDatasetConverter(cache_file_path)
+    if cache_dir is None:
+        cache_dir = SparkSession.builder.getOrCreate().conf \
+            .get("spark.petastorm.converter.default.cache.dir", DEFAULT_CACHE_DIR)
+    dataset_size = df.count()
```

Comment:
There is a performance issue here:

The spark dataframe may be computed twice here: the first for `count()` and the second for materialization.

To optimize it, you can do:
1) materialize spark dataframe to disk first.
2) do "count" via "spark.read.parquet({df_cache_path}).count()", note this will be fast because parquet will store count info in each row groups.